src/yenc.cc

```diff
@@ -23,225 +23,1053 @@
-    unsigned long long part_end = 0;
-    unsigned long long part_size = 0;
-    const char* crc_pos;
-
-    // Verify it's a bytearray
-    if (!PyMemoryView_Check(Py_memoryview_obj)) {
-        PyErr_SetString(PyExc_TypeError, "Expected memoryview");
-        return NULL;
+/**
+ * Parse up to 64 bit representations of a CRC32 hash, discarding the upper 32 bits.
+ * This is necessary because some posts have malformed hashes that exceed 32 bits.
+ * 
+ * @param crc32 String view containing hexadecimal CRC value
+ * @return Optional uint32_t with the parsed CRC, or nullopt if parsing fails
+ */
+std::optional<uint32_t> parse_crc32(std::string_view crc32) {
+    uint64_t value = 0;
+    auto [ptr, ec] = std::from_chars(crc32.data(), crc32.data() + crc32.size(), value, 16);
+
+    // Fail if conversion failed
+    if (ec != std::errc{}) {
+        return std::nullopt;
     }
 
-    // Get buffer and check it is a valid size and type
-    Py_buffer_obj = PyMemoryView_GET_BUFFER(Py_memoryview_obj);
-    if (!PyBuffer_IsContiguous(Py_buffer_obj, 'C') || Py_buffer_obj->len <= 0) {
-        PyErr_SetString(PyExc_ValueError, "Invalid data length or order");
-        retval = NULL;
-        goto finish;
+    return static_cast<uint32_t>(value); // Discard upper 32 bits
+}
+
+/**
+ * Decode a string to Python Unicode with encoding fallback.
+ * 
+ * Try UTF-8 first, fall back to Latin-1 for legacy encodings.
+ * 
+ * This handles filenames and other metadata in yEnc headers that may use
+ * non-UTF-8 encodings from older Usenet posts. Latin-1 is used as fallback
+ * because it can decode any byte sequence without error.
+ * 
+ * @param line String view containing the text to decode
+ * @return PyObject* Unicode string, or nullptr if line is empty or all decoding attempts fail
+ * 
+ * Note: Clears Python errors internally when decoding fails.
+ */
+static PyObject* decode_utf8_with_fallback(std::string_view line) {
+    if (line.empty()) return nullptr; // Ignore empty lines
+
+    auto try_decode = [&](auto decoder, const char* errors = nullptr) -> PyObject* {
+        PyObject* result = decoder(line.data(), line.size(), errors);
+        if (!result) PyErr_Clear();
+        return result;
+    };
+
+    PyObject* py_str = try_decode(PyUnicode_DecodeUTF8);
+    if (!py_str) {
+        py_str = try_decode(PyUnicode_DecodeLatin1, "replace");
+    }
+
+    return py_str;
+}
+
+/**
+ * Detect the encoding format of the article by examining a line.
+ * Detects yEnc format (lines starting with "=ybegin ") and UUEncode format
+ * (60/61 character lines starting with 'M', or "begin " header with octal permissions).
+ * 
+ * @param instance The Decoder instance to update
+ * @param line The line to examine for format detection
+ */
+static inline void decoder_detect_format(Decoder* instance, std::string_view line) {
+    // YEnc detection
+	if (starts_with(line, "=ybegin "))
+	{
+	    Py_XDECREF(instance->format);
+		instance->format = ENCODING_FORMAT_YENC;
```

Comment:
Why do we need the reference counting logic? It wil always already have count>0 because of the reference in the `yenc_init`?